vue/sbc-common-components/src/services/token.services.ts

```diff
@@ -0,0 +1,119 @@
+import Keycloak, { KeycloakInitOptions, KeycloakInstance } from 'keycloak-js'
+
+class TokenServices {
+  private kc: KeycloakInstance
+  private static instance: TokenServices
+  private kcLoaded
+  private counter = 0
+  private timerId
+  private kcOptions: KeycloakInitOptions = {
+    onLoad: 'login-required',
+    checkLoginIframe: false,
+    timeSkew: 0,
+    token: sessionStorage.getItem('KEYCLOAK_TOKEN'),
+    refreshToken: sessionStorage.getItem('KEYCLOAK_REFRESH_TOKEN'),
+    idToken: sessionStorage.getItem('KEYCLOAK_ID_TOKEN')
+  }
+
+  initUsingKc (kcInstance: KeycloakInstance) {
+    console.info('[TokenServices] KC Instance registered using kc instance')
+    this.kc = kcInstance
+  }
+
+  initUsingUrl (keyCloakConfigurl: string) {
+    console.info('[TokenServices] KC Instance registered using kc URL')
+    var self = this
+    return new Promise((resolve, reject) => {
+      this.kc = Keycloak(keyCloakConfigurl)
+      this.kc.init(this.kcOptions
+      ).success(function (authenticated) {
+        console.info('[TokenServices] is User Authenticated:Syncing' + authenticated)
+        self.syncSessionStorage()
+        resolve(self.kc.token)
+      }).error(function (err) {
+        console.info('[TokenServices] Fatal Error:Cudnt Initialise KC instance' + err)
+        reject(new Error('Cudnt Initialise KC'))
+      })
+    })
+  }
+
+  scheduleRefreshTimer () {
+    console.info('[TokenServices Starting the timer] ')
+    this.scheduleRefreshToken()
+  }
+
+  refreshToken () {
+    console.log('[TokenServices] One time Token Refreshing ')
+    return new Promise((resolve, reject) => {
+      this.kc.updateToken(-1)
+        .success(refreshed => {
+          if (refreshed) {
+            console.log('[TokenServices] One time Token Refreshed ')
+            this.syncSessionStorage()
+            resolve()
+          }
+        })
+        .error(() => {
+          // this.cleanupSession()
+          reject(new Error('Cudnt refresh Token'))
+        })
+    })
+  }
+
+  stopRefreshTimer () {
+    console.info('[TokenServices Stopping the timer] ')
+    clearTimeout(this.timerId)
+  }
+
+  private scheduleRefreshToken () {
+    let self = this
+
+    // check if refresh token is still valid . Or else clear all timers and throw errors
+    let refreshTokenExpiresIn = self.kc.refreshToken['exp'] - Math.ceil(new Date().getTime() / 1000) + self.kc.timeSkew
```

Comment:
`self.kc.refreshToken['exp']` is initially undefined, so `refreshTokenExpiresIn` is NaN.

Lucky for you `NaN < 0` is false so the next block of code is not executed.

But there is a logic problem here that should be fixed.

Reply:
self.kc.refreshToken['exp'] shouldnt be Nan when scheduleRefreshToken is invoked. the no token check will be in init. The the ['exp'] will never be undefined.

Reply:
self.kc.refreshToken['exp'] shouldnt be Nan when scheduleRefreshToken is invoked. the no token check will be in init. The the ['exp'] will never be undefined
Whats the logic problem?

Reply:
I found that when I schedule a timer refresh (right after first init), kc does not contain property 'exp' (it's undefined). So the math above fails with NaN.

Am I missing an init step? I just copied the code in the readme.

Reply:
I found that when I schedule a timer refresh (right after first init), kc does not contain property 'exp' (it's undefined). So the math above fails with NaN.

Am I missing an init step? I just copied the code in the readme.

Reply:
I fixed the wrong variable ,thanks for noticing it